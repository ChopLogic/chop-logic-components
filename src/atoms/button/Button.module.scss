@use '/src/styles/typography';
@use '/src/styles/mixins';

.extended {
  width: 100% !important;
}

//.primary {
//  position: relative;
//  border: none;
//  background: transparent;
//  padding: 0;
//  cursor: pointer;
//  outline: none;
//  transition: filter 250ms;
//  width: inherit;
//
//  &__extended {
//    width: 100%;
//  }
//
//  &[disabled] {
//    @include mixins.disabled();
//  }
//
//  &_shadow {
//    position: absolute;
//    top: 0;
//    left: 0;
//    width: 100%;
//    height: 100%;
//    border-radius: var(--cl-border-radius);
//    background: hsl(0deg 0% 0% / 0.25);
//    will-change: transform;
//    transform: translateY(2px);
//    transition: transform 600ms cubic-bezier(0.3, 0.7, 0.4, 1);
//  }
//
//  &_edge {
//    position: absolute;
//    top: 0;
//    left: 0;
//    width: 100%;
//    height: 100%;
//    border-radius: var(--cl-border-radius);
//    background: var(--cl-accent-gradient);
//  }
//
//  &_front {
//    display: flex;
//    align-items: center;
//    justify-content: center;
//    gap: var(--cl-s-gap);
//    position: relative;
//    padding: var(--cl-m-gap) var(--cl-l-gap);
//    border-radius: var(--cl-border-radius);
//    font-size: var(--cl-base-font-size);
//    color: #ffffff;
//    background: var(--cl-accent-color);
//    will-change: transform;
//    transform: translateY(-4px);
//    transition: transform 600ms cubic-bezier(0.3, 0.7, 0.4, 1);
//  }
//
//  &:hover {
//    filter: brightness(110%);
//  }
//
//  &:hover &_front {
//    transform: translateY(-6px);
//    transition: transform 250ms cubic-bezier(0.3, 0.7, 0.4, 1.5);
//  }
//
//  &:active &_front {
//    transform: translateY(-2px);
//    transition: transform 34ms;
//  }
//
//  &:hover &_shadow {
//    transform: translateY(4px);
//    transition: transform 250ms cubic-bezier(0.3, 0.7, 0.4, 1.5);
//  }
//
//  &:active &_shadow {
//    transform: translateY(1px);
//    transition: transform 34ms;
//  }
//
//  &:focus-visible {
//    outline: revert;
//    outline: var(--cl-outline-border);
//    outline-offset: 4px;
//  }
//}

.flat {
  @include typography.font-style(core, base);
  color: var(--cl-primary-color);
  border: none;
  cursor: pointer;
  display: flex;
  align-items: center;
  justify-content: center;
  gap: var(--cl-s-gap);
  padding: var(--cl-m-gap) var(--cl-l-gap);
  border-radius: var(--cl-border-radius);
  background: transparent;

  &__extended {
    width: 100%;
  }

  &:hover,
  &:active {
    background: var(--cl-highlight-background);
    transition: all 0.3s ease 0s;
    color: var(--cl-accent-color);
  }

  &:focus-visible {
    outline: var(--cl-outline-border);
  }

  &[disabled] {
    @include mixins.disabled();
  }
}

.icon {
  @include typography.font-style(core, base);
  color: var(--cl-primary-color);
  border: none;
  cursor: pointer;
  display: flex;
  align-items: center;
  justify-content: center;
  padding: var(--cl-m-gap);
  border-radius: var(--cl-border-radius);
  background: transparent;

  &:hover {
    background-color: var(--cl-highlight-background);
    color: var(--cl-accent-color);
    transition: all 0.3s ease 0s;
  }

  &:focus-visible {
    outline: var(--cl-outline-border);
  }

  &[disabled] {
    pointer-events: none;
    opacity: 0.5;
  }

  svg {
    height: var(--cl-icon-size);
    width: var(--cl-icon-size);
  }
}
