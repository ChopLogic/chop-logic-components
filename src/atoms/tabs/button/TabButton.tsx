import { ChopLogicOrientationMode } from '@enums';
import { getClassName } from '@utils';
import React from 'react';

import styles from './TabButton.module.scss';

type Props = {
  title: string;
  tabId: string;
  onTabSelect: (id: string) => void;
  tabPanelId: string;
  isSelected: boolean;
  mode: ChopLogicOrientationMode;
  isDisabled?: boolean;
};

export const TabButton: React.FC<Props> = ({ title, onTabSelect, tabId, isSelected, isDisabled = false, tabPanelId, mode }) => {
  const buttonClass = getClassName([
    styles.tabButton,
    {
      [styles.tabButton__disabled]: isDisabled,
      [styles.tabButton__vertical]: mode === ChopLogicOrientationMode.Vertical,
      [styles.tabButton__selected]: isSelected,
    },
  ]);

  return (
    <button
      role='tab'
      id={tabId}
      onClick={() => onTabSelect(tabId)}
      aria-selected={isSelected}
      aria-controls={tabPanelId}
      className={buttonClass}
    >
      {title}
    </button>
  );
};
