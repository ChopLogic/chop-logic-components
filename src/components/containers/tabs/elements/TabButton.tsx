import React from 'react';

import { ChopLogicTabsMode } from '@/types';

import { StyledTabButton } from '../Tabs.styled';

type ChopLogicTabButtonProps = {
  title: string;
  tabId: string;
  onTabSelect: (id: string) => void;
  tabPanelId: string;
  isSelected: boolean;
  mode: ChopLogicTabsMode;
  isDisabled?: boolean;
};

const ChopLogicTabButton: React.FC<ChopLogicTabButtonProps> = ({
  title,
  onTabSelect,
  tabId,
  isSelected,
  isDisabled = false,
  tabPanelId,
  mode,
}) => {
  return (
    <StyledTabButton
      role='tab'
      id={tabId}
      onClick={() => onTabSelect(tabId)}
      aria-selected={isSelected}
      aria-controls={tabPanelId}
      $selected={isSelected}
      $disabled={isDisabled}
      $mode={mode}
    >
      {title}
    </StyledTabButton>
  );
};

export default ChopLogicTabButton;
