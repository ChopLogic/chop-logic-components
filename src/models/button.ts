import type { ButtonView, IconName } from '@enums';
import type { FocusEventHandler, MouseEventHandler } from 'react';

import type { ChopLogicInputProps } from './_common';

export interface ButtonProps extends Partial<ChopLogicInputProps> {
  onClick?: MouseEventHandler<HTMLButtonElement>;
  onFocus?: FocusEventHandler<HTMLButtonElement>;
  onBlur?: FocusEventHandler<HTMLButtonElement>;
  view?: ButtonView;
  icon?: IconName;
  text?: string;
  extended?: boolean;
  type?: 'button' | 'reset' | 'submit';
  tooltip?: string;
}
