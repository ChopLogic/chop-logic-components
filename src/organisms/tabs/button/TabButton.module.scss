@use '/src/styles/typography';
@use '/src/styles/mixins';

.tab {
  @include typography.font-style(core, base);

  color: var(--cl-primary-color);

  background: var(--cl-primary-background);
  outline: none;
  border: none;
  padding: var(--cl-m-gap);
  cursor: pointer;
  overflow: hidden;
  flex: 1;
  border-radius: var(--cl-border-radius);
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;

  &:focus-visible {
    outline: var(--cl-outline-border);
    outline-offset: -4px;
  }

  &_wrapper {
    position: relative;
    display: flex;
    align-items: center;
    justify-content: space-between;
    gap: var(--cl-l-gap);
    border-top-left-radius: var(--cl-border-radius);
    border-top-right-radius: var(--cl-border-radius);
    border-top: var(--cl-primary-border);
    border-right: var(--cl-primary-border);
    border-left: var(--cl-primary-border);
    background-color: var(--cl-primary-background);

    &__selected {
      top: 1px;
    }

    &__vertical {
      border-bottom-left-radius: var(--cl-border-radius);
      border-top-right-radius: 0;
      border-top: var(--cl-primary-border);
      border-right: none;
      border-left: var(--cl-primary-border);
      border-bottom: var(--cl-primary-border);

      &.tab_wrapper__selected {
        top: 0;
        left: 1px;
      }
    }

    &__stretched {
      flex: 1;
    }

    &__disabled {
      @include mixins.disabled();
    }

    &__inactive {
      opacity: 0.7;

      &:hover {
        opacity: 1;
      }
    }
  }

  &_editButton {
    margin-right: var(--cl-s-gap);
  }
}
