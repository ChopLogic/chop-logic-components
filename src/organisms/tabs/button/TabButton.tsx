import { Button } from '@atoms';
import { ButtonView, IconName, OrientationMode } from '@enums';
import { getClassName } from '@utils';
import { FC, useEffect, useId, useRef, useState } from 'react';

import { TabEditInput } from '../edit-input/TabEditInput';
import styles from './TabButton.module.scss';

type Props = {
  title: string;
  tabId: string;
  onTabSelect: (id: string) => void;
  tabPanelId: string;
  isSelected: boolean;
  mode: OrientationMode;
  isDisabled?: boolean;
  stretched?: boolean;
  editable?: boolean;
  onTabTitleChange?: (newTitle: string) => void;
  onTabDelete?: (id: string) => void;
  extendable?: boolean;
};

export const TabButton: FC<Props> = ({
  title,
  onTabSelect,
  onTabTitleChange,
  onTabDelete,
  tabId,
  isSelected,
  isDisabled = false,
  stretched = false,
  editable = false,
  extendable = false,
  tabPanelId,
  mode,
}) => {
  const [isEditMode, setIsEditMode] = useState(false);
  const [editValue, setEditValue] = useState(title);
  const inputRef = useRef<HTMLInputElement>(null);
  const cancelButtonId = useId();
  const isEditToggleVisible = editable && !isDisabled && isSelected;
  const isDeleteButtonVisible = extendable && !isDisabled && !!onTabDelete && isSelected;
  const hasAdditionalButtons = isEditToggleVisible || isDeleteButtonVisible;

  const wrapperClass = getClassName([
    styles.tab_wrapper,
    {
      [styles.tab_wrapper__disabled]: isDisabled,
      [styles.tab_wrapper__vertical]: mode === OrientationMode.Vertical,
      [styles.tab_wrapper__selected]: isSelected,
      [styles.tab_wrapper__stretched]: stretched,
      [styles.tab_wrapper__editable]: editable,
      [styles.tab_wrapper__inactive]: !isSelected,
    },
  ]);

  // Focus input when entering edit mode
  useEffect(() => {
    if (isEditMode && inputRef.current) {
      inputRef.current.focus();
      inputRef.current.select();
    }
  }, [isEditMode]);

  // Sync edit value with title prop
  useEffect(() => {
    setEditValue(title);
  }, [title]);

  const toggleEditMode = () => {
    if (isDisabled) return;
    setIsEditMode(!isEditMode);
  };

  const handleTabClick = () => {
    if (isEditMode) return; // Don't select tab while editing
    onTabSelect(tabId);
  };

  const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setEditValue(e.target.value);
  };

  const handleInputBlur = (e: React.FocusEvent<HTMLInputElement>) => {
    // Check if the blur was caused by clicking on a cancel button
    const relatedTarget = e.relatedTarget as HTMLElement;
    const isBlurCausedByButton = relatedTarget?.id === cancelButtonId;

    // Only call onInputBlur if the blur wasn't caused by a button click
    if (!isBlurCausedByButton) {
      saveTitleChange();
    }
  };

  const handleInputKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (e.key === 'Enter') {
      e.preventDefault();
      saveTitleChange();
    } else if (e.key === 'Escape') {
      e.preventDefault();
      cancelEdit();
    }
  };

  const saveTitleChange = () => {
    if (editValue.trim() && editValue !== title) {
      onTabTitleChange?.(editValue.trim());
    } else {
      // Reset to original title if empty or unchanged
      setEditValue(title);
    }
    setIsEditMode(false);
  };

  const cancelEdit = () => {
    setEditValue(title);
    setIsEditMode(false);
  };

  return (
    <span className={wrapperClass}>
      {isEditMode ? (
        <TabEditInput
          tabId={tabId}
          editValue={editValue}
          isEditMode={isEditMode}
          onInputChange={handleInputChange}
          onInputBlur={handleInputBlur}
          onInputKeyDown={handleInputKeyDown}
          onSave={saveTitleChange}
          onCancel={cancelEdit}
          cancelButtonId={cancelButtonId}
        />
      ) : (
        <>
          <button
            role='tab'
            id={tabId}
            onClick={handleTabClick}
            aria-selected={isSelected}
            aria-controls={tabPanelId}
            className={styles.tab}
            disabled={isDisabled}
          >
            {editValue}
          </button>
          {hasAdditionalButtons && (
            <span className={styles.tab_buttons}>
              {isEditToggleVisible && (
                <Button view={ButtonView.Icon} icon={IconName.Edit} onClick={toggleEditMode} label='Edit tab' tooltip='Edit tab' />
              )}
              {isDeleteButtonVisible && (
                <Button
                  view={ButtonView.Icon}
                  icon={IconName.Delete}
                  onClick={() => onTabDelete?.(tabId)}
                  label='Delete tab'
                  tooltip='Delete tab'
                />
              )}
            </span>
          )}
        </>
      )}
    </span>
  );
};
