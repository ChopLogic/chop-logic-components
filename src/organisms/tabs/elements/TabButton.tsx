import React from 'react';
import { ChopLogicOrientationMode } from '@enums';
import { getClassName } from '@utils';
import styles from '../Tabs.module.scss';

type ChopLogicTabButtonProps = {
  title: string;
  tabId: string;
  onTabSelect: (id: string) => void;
  tabPanelId: string;
  isSelected: boolean;
  mode: ChopLogicOrientationMode;
  isDisabled?: boolean;
};

const ChopLogicTabButton: React.FC<ChopLogicTabButtonProps> = ({
  title,
  onTabSelect,
  tabId,
  isSelected,
  isDisabled = false,
  tabPanelId,
  mode,
}) => {
  const buttonClass = getClassName([
    styles.tabButton,
    {
      [styles.tabButton__disabled]: isDisabled,
      [styles.tabButton__vertical]: mode === ChopLogicOrientationMode.Vertical,
      [styles.tabButton__selected]: isSelected,
    },
  ]);

  return (
    <button
      role='tab'
      id={tabId}
      onClick={() => onTabSelect(tabId)}
      aria-selected={isSelected}
      aria-controls={tabPanelId}
      className={buttonClass}
    >
      {title}
    </button>
  );
};

export default ChopLogicTabButton;
