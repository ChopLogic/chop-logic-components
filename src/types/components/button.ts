import { MouseEventHandler } from 'react';

import { ChopLogicButtonView, ChopLogicIconName } from '@/enums';

import { CommonInputProps } from './_common';

export type ChopLogicButtonProps = Partial<CommonInputProps> & {
  onClick?: MouseEventHandler<HTMLButtonElement>;
  view?: ChopLogicButtonView;
  icon?: ChopLogicIconName;
  text?: string;
  extended?: boolean;
  type?: 'button' | 'reset' | 'submit';
  tooltip?: string;
};
