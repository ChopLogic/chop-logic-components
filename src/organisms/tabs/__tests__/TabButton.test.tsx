import { OrientationMode } from '@enums';
import { render, screen } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';

import { TabButton } from '../__elements__/button/TabButton';

describe('TabButton', () => {
  const testProps = {
    title: 'Tab button',
    tabId: 'id-1',
    onTabSelect: vi.fn(),
    tabPanelId: 'panel-id-1',
    isSelected: true,
    mode: OrientationMode.Vertical,
    isDisabled: false,
  };

  it('should match the snapshot', () => {
    const { asFragment } = render(<TabButton {...testProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should have the aria attribute', () => {
    render(<TabButton {...testProps} />);
    expect(screen.getByRole('tab')).toBeInTheDocument();
  });
});
