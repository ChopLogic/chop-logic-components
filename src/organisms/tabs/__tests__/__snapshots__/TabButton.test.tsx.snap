// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TabButton > should match the snapshot 1`] = `
<DocumentFragment>
  <button
    aria-controls="panel-id-1"
    aria-selected="true"
    class="_tabButton_54926d _tabButton__vertical_54926d _tabButton__selected_54926d"
    id="id-1"
    role="tab"
  >
    Tab button
  </button>
</DocumentFragment>
`;
